In [4]:
import numpy as np
import pandas as pd
  
# read text file into pandas DataFrame and
# create header
df = pd.read_csv("adaptive-e-learning-for-educational-recommendation-system/dataset/transactions_v2.txt", dtype=str ,sep="\t", header=None, names=['timestamp','course','student','resourceID','partID','mode','result'])
  

In [3]:
# df=pd.read_csv('adaptive-e-learning-for-educational-recommendation-system/dataset/Loncapa.csv',low_memory=False) 

In [5]:
df

timestamp         course        student  \
0         1292676876  17PfVQUjnFksY  42xcn1otKzbyA   
1         1292676810  17PfVQUjnFksY  42G67UH9o9VC.   
2         1292676473  17PfVQUjnFksY  42nnWozrfbHWE   
3         1292676458  17PfVQUjnFksY  42nnWozrfbHWE   
4         1292676405  17PfVQUjnFksY  42twHEdc5L6WM   
...              ...            ...            ...   
29306635  1231167493  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306636  1231167493  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306637  1231167493  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306638  1231167394  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306639  1231167328  17lWZ7mTKeQg.  42.SSHZ8gHchE   

                                                 resourceID partID mode result  
0                          lib/templates/examupload.problem    NaN  NaN    NaN  
1         ext/www.pa.msu.edu/~osheabr/teaching/LB273_gra...    NaN  NaN    NaN  
2         msu/kortemey/physicslib/thermo/tollbooth2.problem    NaN  NaN    NaN  
3         msu/kortemey/physicslib/oscillations/resonance...    NaN  NaN    NaN  
4                          lib/templates/examupload.problem    NaN  NaN    NaN  
...                                                     ...    ...  ...    ...  
29306635  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306636  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306637  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306638  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306639                msu/nagytibo/Misc/clickregA.problem    NaN  NaN    NaN  

[29306640 rows x 7 columns]

In [12]:
data=df.to_numpy()
data[:,1]

array(['17PfVQUjnFksY', '17PfVQUjnFksY', '17PfVQUjnFksY', ...,
       '17lWZ7mTKeQg.', '17lWZ7mTKeQg.', '17lWZ7mTKeQg.'], dtype=object)

In [9]:
course=data[:,1]
np.unique(course)

array(['17.KPLV7108Q.', '170Rg65URrGNE', '1734gygYlAFOI', '173XHazC4puKI',
       '178UapjfFklEc', '179Ce1976MyiM', '17A7imT7M3wxI', '17AlNB3stJY5Q',
       '17DtnibRFokt6', '17EbEJfGNfRzc', '17FGvfsKTd9qg', '17Faxuol3IIFQ',
       '17ICS6N0eBevM', '17IwT5b39kY6Q', '17Np8nZS6oVSc', '17PfVQUjnFksY',
       '17Q.FdNAmYJSE', '17QDnYbcwatec', '17X9ElKlKxuE2', '17XyLE6fAS21U',
       '17b1KWZk/9bWg', '17dSDs2HzHQTU', '17drv4WFUwXr.', '17eUp22S2crNE',
       '17ipAbYEhwvus', '17klvCknkhcmE', '17lNXP7woERBI', '17lWZ7mTKeQg.',
       '17mSq8Ta3yYMs', '17oIHrAYo4YwU', '17oPA6FLPK2B.', '17pHpKo0nAoak',
       '17vKzihe/S1K2', '17y.VYL4BL2JI'], dtype=object)

In [6]:
# save into 34 different .csv files based on 34 different courses
grouped = df.groupby('course')

# Step 2: Create separate arrays for each group.
array_group = [grouped.get_group(name) for name in grouped.groups]

# 'array_group' will be a list containing separate arrays for each unique element in the 'column_name' column.
# Each array will contain the rows of the original DataFrame corresponding to that unique element.

# Example: Accessing the first array for the first unique element ('A'):
array_group
len(array_group)

34

In [67]:
# course_4=course_4.astype(int)
grouped.count()


timestamp  student  resourceID  partID    mode  result
course                                                               
17.KPLV7108Q.     499293   499293      499293   75445   17616  120714
170Rg65URrGNE     792816   792816      792816  199140   21704  266393
1734gygYlAFOI     879669   879669      879669  162528   59572  223557
173XHazC4puKI     825393   825393      825393  211423   28287  281038
178UapjfFklEc    1181936  1181936     1181936  292728     123  409286
179Ce1976MyiM     878915   878915      878915  224710   30182  292500
17A7imT7M3wxI     733077   733077      733077  110739   25262  177600
17AlNB3stJY5Q     166482   166482      166482   33763    4446   40237
17DtnibRFokt6     840995   840995      840995  272206   82469  306045
17EbEJfGNfRzc    1498928  1498928     1498928  459914   82809  493385
17FGvfsKTd9qg     512412   512412      512412  102241   19562  126312
17Faxuol3IIFQ     848691   848691      848691  211385   31057  252379
17ICS6N0eBevM     868847   868847      868847  225112   45646  289716
17IwT5b39kY6Q     926136   926136      926136  246316   34037  294399
17Np8nZS6oVSc     123007   123007      123007   34071    3214   41095
17PfVQUjnFksY     416458   416458      416458   99388   18851  126084
17Q.FdNAmYJSE     920376   920376      920376  174802   44255  229948
17QDnYbcwatec    1087759  1087759     1087759  267358  101122  378034
17X9ElKlKxuE2     459448   459448      459448   82238       4   82073
17XyLE6fAS21U     791316   791316      791316  222348   24932  251341
17b1KWZk/9bWg    1980665  1980665     1980665  461436   90512  647155
17dSDs2HzHQTU     466154   466154      466154   70360   14465  113668
17drv4WFUwXr.     678487   678487      678487  156880   23669  213641
17eUp22S2crNE     683217   683217      683217  132849   40907  160392
17ipAbYEhwvus    1160247  1160247     1160247  243897   60896  402608
17klvCknkhcmE    2084558  2084558     2084558  479511   98209  667242
17lNXP7woERBI    1202331  1202331     1202331  232213   66917  306815
17lWZ7mTKeQg.    1298595  1298595     1298595  267260   57576  411446
17mSq8Ta3yYMs     785318   785318      785318  233382   30560  255322
17oIHrAYo4YwU     689567   689567      689567  124243   41992  169639
17oPA6FLPK2B.    1010569  1010569     1010569  171441   36926  210760
17pHpKo0nAoak     524027   524027      524027  107246   17452  130409
17vKzihe/S1K2    1200160  1200160     1200160  328466   80298  407561
17y.VYL4BL2JI     290791   290791      290791   53818       8   53809

 Eg how often are resources shared across the same course? If we have two similar courses, can we just pool them together? Can we just pool all the courses together even if there isn’t much resource overlap (there is still some common information in how student's generally learn)? 
 
 A total of 9,349 individual students accessed 2,464 different content pages and 3,703 different online problems. Unless we really look into details of each problems to see if those are essentially similar problems, we have too many questions (unknown parameters)

In [4]:
# how often are resources shared across the same course?
# course_overlap=np.zeros((34,34))
# exam_overlap=np.zeros((34,34))
# hw_overlap=np.zeros((34,34))
hw_ques_overlap=np.zeros((34,34))
for i in range(len(array_group)):
    coursei=array_group[i]
    #exami=coursei[coursei['mode']=='exam']
    #handi=coursei[coursei['mode']=='hand']
    hwi=coursei[coursei['mode'].isnull()]
    hwi_ques=hwi[hwi['result'].notnull()]
    hwi_material=hwi[hwi['result'].isnull()]
    #s_i=coursei.to_numpy()[:,2] #students
    #q_i=coursei.to_numpy()[:,3] #questions
    #s_exam_i=exami.to_numpy()[:,2] #students
    #q_exam_i=exami.to_numpy()[:,3] #questions
    #s_hw_i=hwi.to_numpy()[:,2] #students
    q_hw_i=hwi.to_numpy()[:,3] #questions
    q_hw_ques_i=np.unique(hwi_ques.to_numpy()[:,3])
    for j in range(len(array_group)):
        coursej=array_group[j]
        #examj=coursej[coursej['mode']=='exam']
        #handj=coursej[coursej['mode']=='hand']
        hwj=coursej[coursej['mode'].isnull()]
        hwj_ques=hwj[hwj['result'].notnull()]
        hwj_material=hwj[hwj['result'].isnull()]
        #s_j=coursej.to_numpy()[:,2] #students
        #q_j=coursej.to_numpy()[:,3] #questions
        #course_overlap[i,j]=len(np.intersect1d(q_i, q_j))
        #s_exam_j=examj.to_numpy()[:,2] #students
        #q_exam_j=examj.to_numpy()[:,3] #questions
        #s_hw_j=hwj.to_numpy()[:,2] #students
        q_hw_j=hwj.to_numpy()[:,3] #questions
        q_hw_ques_j=np.unique(hwj_ques.to_numpy()[:,3])
        #exam_overlap[i,j]=len(np.intersect1d(q_exam_i, q_exam_j))
        #hw_overlap[i,j]=len(np.intersect1d(q_hw_i, q_hw_j))
        hw_ques_overlap[i,j]=len(np.intersect1d(q_hw_ques_i, q_hw_ques_j))

In [68]:
course_overlap = np.loadtxt('course_overlap.csv')
exam_overlap = np.loadtxt('exam_overlap.csv')
hw_overlap = np.loadtxt('hw_overlap.csv')
hw_ques_overlap = np.loadtxt('hw_ques_overlap.csv')

In [14]:
# course_overlap=course_overlap.astype(int)
# exam_overlap=exam_overlap.astype(int)
# hw_overlap=hw_overlap.astype(int)
hw_ques_overlap=hw_ques_overlap.astype(int)
np.savetxt('hw_ques_overlap.csv',hw_ques_overlap)

In [69]:
# Unique question
hw_ques_0=hw_ques_overlap[0,:]
sort = np.argsort(-hw_ques_0)
# Sort the array in descending order
hw_ques_0_sort = hw_ques_0[sort]
print(sort)
hw_ques_0_sort

[ 0 21  6 18 33  7 10 31 23 30 26 16 17 12  4 20 25 11 13 32 24 27 29 28
  5  1  2 15 14  9  8  3 22 19]


array([214., 211., 211., 196., 166., 135., 135., 134., 122.,  94.,  75.,
        75.,  38.,  36.,  27.,  15.,  15.,  14.,   8.,   1.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.])

In [70]:
# Total question + material
hw_0=hw_overlap[0,:]
sorted_indices = np.argsort(-hw_0)
# Sort the array in descending order
hw_0_sort = hw_0[sorted_indices]
hw_0_sort
print(sorted_indices)
hw_0_sort

[ 0  6 21 10 18 33 25  7 31 23 30  4 20 16 26 29  2 17 12 11 13 15 32  8
  9 28 27 14 24 22  1  5  3 19]


array([625., 607., 575., 556., 541., 531., 520., 494., 494., 450., 417.,
       381., 365., 286., 283., 208., 150.,  45.,  43.,  15.,  12.,   3.,
         2.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.])

In [111]:
hw_overlap[4,:]

array([381,   0, 145,   0, 594,   0, 387, 394,   0,   0, 405,  21,  14,
        25,   0,   1, 221,  16, 378,   0, 566, 384,   0, 352,   8, 572,
       222,   7,   0, 203, 315, 399,  71, 375])

In [46]:
df

timestamp         course        student  \
0         1292676876  17PfVQUjnFksY  42xcn1otKzbyA   
1         1292676810  17PfVQUjnFksY  42G67UH9o9VC.   
2         1292676473  17PfVQUjnFksY  42nnWozrfbHWE   
3         1292676458  17PfVQUjnFksY  42nnWozrfbHWE   
4         1292676405  17PfVQUjnFksY  42twHEdc5L6WM   
...              ...            ...            ...   
29306635  1231167493  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306636  1231167493  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306637  1231167493  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306638  1231167394  17lWZ7mTKeQg.  42.SSHZ8gHchE   
29306639  1231167328  17lWZ7mTKeQg.  42.SSHZ8gHchE   

                                                 resourceID partID mode result  
0                          lib/templates/examupload.problem    NaN  NaN    NaN  
1         ext/www.pa.msu.edu/~osheabr/teaching/LB273_gra...    NaN  NaN    NaN  
2         msu/kortemey/physicslib/thermo/tollbooth2.problem    NaN  NaN    NaN  
3         msu/kortemey/physicslib/oscillations/resonance...    NaN  NaN    NaN  
4                          lib/templates/examupload.problem    NaN  NaN    NaN  
...                                                     ...    ...  ...    ...  
29306635  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306636  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306637  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306638  msu/physicslib/msuphysicslib/51_ElectricForce/...    NaN  NaN    NaN  
29306639                msu/nagytibo/Misc/clickregA.problem    NaN  NaN    NaN  

[29306640 rows x 7 columns]

# Combine similar courses 

In [106]:
# course0=array_group[0]
# course6=array_group[6]
# course21=array_group[21]
# course10=array_group[10]
# course18=array_group[18]
# course33=array_group[33]
# course31=array_group[31]
# course30=array_group[30]
# course23=array_group[23]
# course4=array_group[4]
# combined_df = pd.concat([course0, course6,course21, course10,course18, course33], axis=0)
# combined_df.reset_index(drop=True, inplace=True)
# combined_df

sem1=np.array([0,6, 21, 10, 18, 33, 25,7, 31, 23, 30,4, 20, 16, 26])
course_sem1=pd.DataFrame()
for i in sem1:
    course=array_group[i]
    course_sem1=pd.concat([course_sem1,course],axis=0)

#course_sem1=pd.concat([array_group[0],array_group[21],array_group[6],array_group[18],array_group[10],array_group[33],array_group[25],array_group[7],array_group[23],array_group[4],,array_group[31],array_group[30],array_group[16],array_group[20],array_group[26]], axis=0)

# course3=array_group[3]
# course1=array_group[1]
# course5=array_group[5]
# course22=array_group[22]
# course19=array_group[19]
# course28=array_group[28]
# combined_df_1 = pd.concat([course1, course3,course5, course22,course19,course28], axis=0)
# combined_df_1.reset_index(drop=True, inplace=True)
# combined_df_1

# course11=array_group[11]
# course13=array_group[13]
# course12=array_group[12]
# course17=array_group[17]
# combined_df_11 = pd.concat([course11, course13,course12, course17], axis=0)
# combined_df_11.reset_index(drop=True, inplace=True)
# combined_df_11

# course4=array_group[4]
# course25=array_group[25]
# course20=array_group[20]
# combined_df_4 = pd.concat([course4, course25,course20], axis=0)
# combined_df_4.reset_index(drop=True, inplace=True)
# combined_df_4

#course8=array_group[8] # bcourse 8 is only related to course19 and course 28, independent of the others


timestamp         course        student  \
0        1320647262  170Rg65URrGNE  42E23hDBkViPo   
1        1320647262  170Rg65URrGNE  42E23hDBkViPo   
2        1320647262  170Rg65URrGNE  42E23hDBkViPo   
3        1320647262  170Rg65URrGNE  42E23hDBkViPo   
4        1320647262  170Rg65URrGNE  42E23hDBkViPo   
...             ...            ...            ...   
4752240  1263241800  17mSq8Ta3yYMs  42/JUL0UvfHRs   
4752241  1263241660  17mSq8Ta3yYMs  42rT2iP0QwOmY   
4752242  1263241402  17mSq8Ta3yYMs  42fp8t0RSbtv6   
4752243  1263213089  17mSq8Ta3yYMs  42naKcYsXXubg   
4752244  3155738400  17mSq8Ta3yYMs  42ogOhXtWRUDY   

                                                resourceID partID mode  \
0        msu/nagytibo/ModernPhysics/SpecialRelativity/0...    NaN  NaN   
1        msu/nagytibo/Optics/WaveOptics/Exam/examprob00...    NaN  NaN   
2        msu/textbook/serway/Chap26/prob2606_OldEdition...    NaN  NaN   
3              msu/textbook/serway/Chap24/prob2468.problem    NaN  NaN   
4              msu/nagytibo/Optics/WaveOptics/003a.problem    NaN  NaN   
...                                                    ...    ...  ...   
4752240               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752241               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752242               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752243               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752244  mcgrawhill/BauerWestfall/1stEdition/Chapter25/...      b  NaN   

                      result  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  
...                      ...  
4752240                  NaN  
4752241                  NaN  
4752242                  NaN  
4752243                  NaN  
4752244  incorrect_attempted  

[4752245 rows x 7 columns]

In [110]:
combined_df_1

timestamp         course        student  \
0        1320647262  170Rg65URrGNE  42E23hDBkViPo   
1        1320647262  170Rg65URrGNE  42E23hDBkViPo   
2        1320647262  170Rg65URrGNE  42E23hDBkViPo   
3        1320647262  170Rg65URrGNE  42E23hDBkViPo   
4        1320647262  170Rg65URrGNE  42E23hDBkViPo   
...             ...            ...            ...   
4752240  1263241800  17mSq8Ta3yYMs  42/JUL0UvfHRs   
4752241  1263241660  17mSq8Ta3yYMs  42rT2iP0QwOmY   
4752242  1263241402  17mSq8Ta3yYMs  42fp8t0RSbtv6   
4752243  1263213089  17mSq8Ta3yYMs  42naKcYsXXubg   
4752244  3155738400  17mSq8Ta3yYMs  42ogOhXtWRUDY   

                                                resourceID partID mode  \
0        msu/nagytibo/ModernPhysics/SpecialRelativity/0...    NaN  NaN   
1        msu/nagytibo/Optics/WaveOptics/Exam/examprob00...    NaN  NaN   
2        msu/textbook/serway/Chap26/prob2606_OldEdition...    NaN  NaN   
3              msu/textbook/serway/Chap24/prob2468.problem    NaN  NaN   
4              msu/nagytibo/Optics/WaveOptics/003a.problem    NaN  NaN   
...                                                    ...    ...  ...   
4752240               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752241               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752242               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752243               msu/mmp/kap18/problems/cd459.problem    NaN  NaN   
4752244  mcgrawhill/BauerWestfall/1stEdition/Chapter25/...      b  NaN   

                      result  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  
...                      ...  
4752240                  NaN  
4752241                  NaN  
4752242                  NaN  
4752243                  NaN  
4752244  incorrect_attempted  

[4752245 rows x 7 columns]

In [114]:
# Can we just pool all the courses together? 
# A total of 9,349 individual students accessed 2,464 different content pages and 3,703 different online problems
exam=course_sem1[course_sem1['mode']=='exam']
hand=course_sem1[course_sem1['mode']=='hand']
hw=course_sem1[course_sem1['mode'].isnull()]

HW_ques=hw[hw['result'].notnull()]
HW_material=hw[hw['result'].isnull()]
S_HW=HW_ques.to_numpy()[:,2]
N_HW=np.unique(S_HW)
print('The number of students N in hw: ',len(N_HW))
Q_HW=np.unique(HW_ques.to_numpy()[:,3])
print('The number of questions Q in hw: ',len(Q_HW))
print(HW_ques['result'].unique())
# sort the df by descending time
HW_ques_sort = HW_ques.sort_values(by='timestamp', ascending=False)
HW_ques_sort = HW_ques_sort[HW_ques_sort['result'] != 'ungraded_attempted']
HW_ques_sort = HW_ques_sort[HW_ques_sort['result'] != 'excused']
HW_ques_sort = HW_ques_sort[HW_ques_sort['result'] != 'inealGases']
HW_ques_sort=HW_ques_sort.reset_index(drop=True)

The number of students N in hw:  5045
The number of questions Q in hw:  1029
['incorrect_attempted' 'correct_by_student' 'ungraded_attempted' 'excused'
 'incorrect_attempt1159889901' 'inealGases']


In [117]:
HW_ques
HW_ques_sort

timestamp         course        student  \
0        1295338017  17FGvfsKTd9qg  42Z2Q5wFk/Xkg   
1        1295338010  17FGvfsKTd9qg  42Z2Q5wFk/Xkg   
2        1286626348  17pHpKo0nAoak  42Z2Q5wFk/Xkg   
3        1286626278  17pHpKo0nAoak  42Z2Q5wFk/Xkg   
4        1286626140  17pHpKo0nAoak  42Z2Q5wFk/Xkg   
...             ...            ...            ...   
2996423  1030372677  17y.VYL4BL2JI  42bee01G4HQN.   
2996424  1030372591  17y.VYL4BL2JI  42bee01G4HQN.   
2996425  1030368937  17y.VYL4BL2JI  42hjm8IFM3D8k   
2996426  1030360245  17y.VYL4BL2JI  424hcgCJ9kPss   
2996427  1029848488  17y.VYL4BL2JI  42wE5lhXnq87.   

                                                resourceID partID mode  \
0            msu/nagytibo/Mechanics/Kinematics/001.problem     02  NaN   
1            msu/nagytibo/Mechanics/Kinematics/001.problem     02  NaN   
2        mcgrawhill/BauerWestfall/1stEdition/Chapter07/...      d  NaN   
3        mcgrawhill/BauerWestfall/1stEdition/Chapter07/...      d  NaN   
4        mcgrawhill/BauerWestfall/1stEdition/Chapter07/...      c  NaN   
...                                                    ...    ...  ...   
2996423                msu/mmp/kap1/problems/cd013.problem      0  NaN   
2996424                msu/mmp/kap1/problems/cd012.problem      0  NaN   
2996425                msu/mmp/kap1/problems/cd012.problem      0  NaN   
2996426                msu/mmp/kap1/problems/cd015.problem      0  NaN   
2996427             msu/mmp/kap2/problems/click_v0.problem      0  NaN   

                      result  
0        incorrect_attempted  
1        incorrect_attempted  
2         correct_by_student  
3        incorrect_attempted  
4         correct_by_student  
...                      ...  
2996423  incorrect_attempted  
2996424   correct_by_student  
2996425   correct_by_student  
2996426   correct_by_student  
2996427  incorrect_attempted  

[2996428 rows x 7 columns]

In [122]:

# HW_ques_sort
# HW_ques_sort['timestamp'] == HW_ques_sort['timestamp'].sort_values(ascending=False)
# a= HW_ques_sort['timestamp'].reset_index(drop=True)
# b = HW_ques_sort['timestamp'].sort_values(ascending=False)
# b.reset_index(drop=True)
# a==b
# is_decreasing = HW_ques_sort['timestamp'].is_monotonic_decreasing
# is_decreasing
HW_ques_sort['result'].unique()

array(['incorrect_attempted', 'correct_by_student',
       'incorrect_attempt1159889901'], dtype=object)

In [123]:
my_dict = {} # dictionary for question id

for i in range(len(Q_HW)):
    my_dict[Q_HW[i]] = i  # You can assign values to the keys if needed

f1=np.frompyfunc(lambda hw_name, dic: dic[hw_name], 2, 1)
# my_dict['msu/hallstei/PHY 231C/Homework/MISC/hw001.problem']
total_array=[]
for n in range(len(N_HW)):
    N=HW_ques_sort[HW_ques_sort['student']==N_HW[n]]
    total_array.append(N.shape[0]) # length
    q_list=N.to_numpy()[:,3][::-1]# we want order according to time
    result_list=N.to_numpy()[:,6][::-1]
    q_id=f1(q_list, my_dict)
    output = ''# string type
    flag_q=True
    for i in q_id:
        if flag_q:
            flag_q = False
        else:    train_result="Standard BKT (train_auc):"+str(model.evaluate(data_path = "data/df1_bkt_train.csv", metric="auc"))
    valid_result="Standard BKT (valid_auc):"+str(model.evaluate(data_path = "data/df1_bkt_test.csv", metric="auc"))
    print(train_result)
    print(valid_result)
    logger.info(train_result)
    logger.info(valid_result)
            output+=','
        output += f"{i}"
    total_array.append(output) # ques
    ans=''
    flag = True
    for i in range(len(result_list)):
        if flag:
            flag = False
        else:
            ans += ','
        if result_list[i]=='correct_by_student':
            ans +=str(1)
        else:
            ans+=str(0)
    total_array.append(ans)

In [126]:
len(total_array)

15135

In [127]:
# if any(element == '' for element in total_array):
#     print("The list contains empty elements.")
# else:
#     print("The list does not contain empty elements.")
# Check and find if there are empty elements in the list 

indices_of_empty_elements = [index for index, element in enumerate(total_array) if element == '']

if indices_of_empty_elements:
    print("Empty elements found at indices:", indices_of_empty_elements)
else:
    print("No empty elements found in the list.")
idx_remove=[4656, 6711,11931]
idx_remove.extend(indices_of_empty_elements)
total_array_update = [element for index, element in enumerate(total_array) if index not in idx_remove]

Empty elements found at indices: [4657, 4658, 6712, 6713, 11932, 11933]


In [128]:
file_path = "sem1_new.csv"
np.savetxt(file_path, total_array_update, delimiter=',', fmt='%s')

In [130]:
sem1_new=pd.read_csv('sem1_new.csv',header=None,sep="\t")
sem1_new
sem1_new_train=sem1_new.to_numpy()[0:12000]
sem1_new_test=sem1_new.to_numpy()[12000:]
np.savetxt('sem1_new_train.csv', sem1_new_train, delimiter=',', fmt='%s')
np.savetxt('sem1_new_test.csv', sem1_new_test, delimiter=',', fmt='%s')

In [133]:
sem1_new_train=pd.read_csv('sem1_new_train.csv',header=None,sep='\t')
sem1_new_train

0
0                                                    737
1      58,58,61,61,82,82,83,83,84,763,763,763,763,763...
2      0,0,0,1,0,1,0,1,0,0,0,0,0,0,0,1,1,1,0,0,0,0,0,...
3                                                    846
4      460,460,480,480,450,483,459,449,449,487,487,48...
...                                                  ...
11995  238,243,242,242,242,242,460,460,480,450,483,48...
11996  1,1,0,0,1,1,0,1,1,1,0,1,0,0,1,1,0,0,1,1,0,1,0,...
11997                                                636
11998  717,752,753,60,754,58,58,58,58,59,59,59,59,61,...
11999  1,1,1,1,1,0,0,0,1,0,0,0,1,0,0,0,0,0,0,1,1,0,1,...

[12000 rows x 1 columns]

In [52]:
# Convert it into BKT dataset form

ques_dict_4 = {} # dictionary for question id
stud_dict_4 = {} # dictionary for student id
length=HW_ques_sort.to_numpy().shape[0]
comb_df_1=np.zeros((length,3))
for i in range(len(Q_HW)):
    ques_dict_4[Q_HW[i]] = i  # You can assign values to the keys if needed
for i in range(len(N_HW)):
    stud_dict_4[N_HW[i]] = i  # You can assign values to the keys if needed
    
f1=np.frompyfunc(lambda hw_name, dic: dic[hw_name], 2, 1)

# # my_dict['msu/hallstei/PHY 231C/Homework/MISC/hw001.problem']
for n in range(length):
    stud=HW_ques_sort.to_numpy()[n,2]
    ques=HW_ques_sort.to_numpy()[n,3]
    result=HW_ques_sort.to_numpy()[n,6]
    q_id=f1(ques, ques_dict_4)
    s_id=f1(stud, stud_dict_4)
    comb_df_1[n,0]=s_id
    comb_df_1[n,1]=q_id
    if result=='correct_by_student':
        comb_df_1[n,2]=1
    else:
        comb_df_1[n,2]=0
course_sem1_bkt=comb_df_1.astype(int)

In [92]:
# file_path = "course_sem1_bkt.csv"
# np.savetxt(file_path, course_sem1_bkt, delimiter=',', fmt='%s')
course4_bkt=pd.read_csv('course4_bkt.csv',header=None,sep='\t',delimiter=',')
course4_bkt
# course_sem1_bkt_train=course_sem1_bkt.to_numpy()[0:2400000]
# course_sem1_bkt_test=course_sem1_bkt.to_numpy()[2400000:]
# np.savetxt('course_sem1_bkt_train.csv', course_sem1_bkt_train, delimiter=',', fmt='%s')
# np.savetxt('course_sem1_bkt_test.csv', course_sem1_bkt_test, delimiter=',', fmt='%s')

0   1  2
0       377  27  0
1       377  27  0
2       778   1  0
3       778   1  0
4       778   1  1
...     ...  .. ..
409158  607  77  0
409159  607  77  0
409160  607  77  0
409161  581  56  1
409162  581  75  1

[409163 rows x 3 columns]

In [56]:
course_sem1_bkt_train=pd.read_csv('course_sem1_bkt_train.csv',header=None,sep='\t',delimiter=',')
course_sem1_bkt_train


0    1  2
0        2901  678  0
1        2901  678  0
2        2901   12  1
3        2901   12  0
4        2901   12  1
...       ...  ... ..
2399995  3022   82  1
2399996  4591   87  1
2399997  1081   86  0
2399998  2010   81  0
2399999  4601   86  0

[2400000 rows x 3 columns]

In [286]:

# for i in range(len(array_group)):
#     course1=array_group[4]
#     exam1=course1[course1['mode']=='exam']
#     hand1=course1[course1['mode']=='hand']
#     hw1=course1[course1['mode'].isnull()]
#     s1=course1.to_numpy()[:,2]
#     N_learner+=np.unique(s1)
#     print('The number of students N: ',len(N_learner))
#     # how many questions in exam
#     q1=exam1.to_numpy()[:,3]
#     Q=np.unique(q1)
s1=df.to_numpy()[:,2]
print(np.unique(s1))
q1=df.to_numpy()[:,3]
print(np.unique(q1))

['42.08atpUHfqE' '42.0QTwpmhB0o' '42.1VqZ6o2roA' ... '42zyAZyYyqfVQ'
 '42zzAbMQFmoxo' 'knupplau%ource%2e0%2etries=1:']
['/res/msu/kashy/physicsLib02/15_CM_MomentInertia/CMptmassgrid.problem'
 'Selwyn/mopez/relativity/Time Dilation/time2.problem'
 "bellaire/tbaeckeroot/Workspace07/Snell's Law.problem" ...
 'wiley/hrw/9e/chapter34/34-010.problem'
 'wiley/hrw/9e/chapter35/35-016.problem'
 'wiley/hrw/9e/chapter35/35-017.problem']


In [288]:
print(len(np.unique(s1)))
print(len(np.unique(q1)))

9374
6167


In [245]:
#print(hw1['result'].notnull().sum())
hw_ques=hw1[hw1['result'].notnull()]
hw_material=hw1[hw1['result'].isnull()]
#exam1[exam1['resourceID']==Q[1]]
hw_ung=hw1[hw1['result']=='ungraded_attempted']
hw_cor=hw1[hw1['result']=='correct_by_student']
hw_incor=hw1[hw1['result']=='incorrect_attempted']
hw1['result'].unique()
# hw1[hw1['result']=='excused']
# # exc['resourceID'].unique()
# hw_cor['partID'].unique()

array([nan, 'correct_by_student', 'incorrect_attempted'], dtype=object)

In [26]:
# See the percentage of observation
no_ans=exam1[exam1['result']=='noanswer_or_invalid'].shape[0]
corr_ans=exam1[exam1['result']=='correct_by_scantron'].shape[0]
wrong_ans=exam1[exam1['result']=='incorrect_attempted'].shape[0]
exam1[exam1['result']=='incorrect_attempted']
print('The percentage of observation is',1-no_ans/exam1.shape[0])

The percentage of observation is 0.49968776610843035


In [82]:
# how many students in exam
s1=exam1.to_numpy()[:,2]
N_learner=np.unique(s1)
print('The number of students N: ',len(N_learner))
# how many questions in exam
q1=exam1.to_numpy()[:,3]
Q=np.unique(q1)
print('The number of questions Q: ',len(Q))

The number of students N:  0
The number of questions Q:  0


# Record dataset
course 0,4,5,6,7,8,9,10,11,12,13,17,19,21,27,28,31,32, homework: array([nan, 'incorrect_attempted', 'correct_by_student'], dtype=object), no ungraded_attempted 
course 1: array([nan, 'correct_by_student', 'incorrect_attempted','ungraded_attempted'], dtype=object)
course 23, 33: array([nan, 'correct_by_student', 'incorrect_attempted', 'excused'], dtype=object)-- just very few entries, one question
course 25: array([nan, 'correct_by_student', 'incorrect_attempted','incorrect_attempt1159889901', 'ungraded_attempted'], dtype=object)-just one data


# Course 0 HW: 189 students and 214 questions
# Course 4 HW: 860 students and 161 questions
 Course 5 HW: 393 students and 207 questions
 Course 6 HW: 292 students and 223 questions
 Course 7 HW: 75 students and 224 questions
 Course 8 HW: 456 students and 138 questions
 Course 9 HW: 483 students and 215 questions
 Course 10 HW: 225 students and 190 questions
 Course 11 HW: 470 students and 160 questions
 Course 12 HW: 329 students and 246 questions
 Course 13 HW: 488 students and 177 questions
 Course 17 HW: 431 students and 246 questions
 Course 19 HW: 314 students and 255 questions
 Course 21 HW: 196 students and 222 questions
 Course 27 HW: 606 students and 193 questions
 Course 28 HW: 430 students and 135 questions
 Course 31 HW: 212 students and 200 questions
 Course 32 HW: 472 students and 219 questions

In [15]:
exam=combined_df_1[combined_df_1['mode']=='exam']
hand=combined_df_1[combined_df_1['mode']=='hand']
hw=combined_df_1[combined_df_1['mode'].isnull()]
HW_ques=hw[hw['result'].notnull()]
HW_material=hw[hw['result'].isnull()]
S_HW=HW_ques.to_numpy()[:,2]
N_HW=np.unique(S_HW)
print('The number of students N in hw: ',len(N_HW))
Q_HW=np.unique(HW_ques.to_numpy()[:,3])
print('The number of questions Q in hw: ',len(Q_HW))
print(HW_ques['result'].unique())
# sort the df by descending time
HW_ques_sort = HW_ques.sort_values(by='timestamp', ascending=False)
HW_ques_sort=HW_ques_sort.reset_index(drop=True)
HW_ques_sort

The number of students N in hw:  2080
The number of questions Q in hw:  493
['correct_by_student' 'incorrect_attempted' 'ungraded_attempted']


timestamp         course        student  \
0        3155738400  17mSq8Ta3yYMs  42ogOhXtWRUDY   
1        1298311300  17XyLE6fAS21U  42t6Y2LkeEt1Y   
2        1298311093  17XyLE6fAS21U  42t6Y2LkeEt1Y   
3        1298310999  17XyLE6fAS21U  42t6Y2LkeEt1Y   
4        1298310906  17XyLE6fAS21U  42t6Y2LkeEt1Y   
...             ...            ...            ...   
1400896  1199539489  173XHazC4puKI  42AAcbCjrFHPI   
1400897  1199539398  173XHazC4puKI  42AAcbCjrFHPI   
1400898  1199539361  173XHazC4puKI  42AAcbCjrFHPI   
1400899  1199538747  173XHazC4puKI  42AAcbCjrFHPI   
1400900  1199538743  173XHazC4puKI  42AAcbCjrFHPI   

                                                resourceID partID mode  \
0        mcgrawhill/BauerWestfall/1stEdition/Chapter25/...      b  NaN   
1         msu/mmp/relativity/problems/relativity09.problem      0  NaN   
2         msu/mmp/relativity/problems/relativity07.problem      0  NaN   
3         msu/mmp/relativity/problems/relativity02.problem      0  NaN   
4         msu/mmp/relativity/problems/relativity01.problem      0  NaN   
...                                                    ...    ...  ...   
1400896              msu/mmp/kap18/problems/cd460b.problem      0  NaN   
1400897              msu/mmp/kap18/problems/cd460a.problem      0  NaN   
1400898               msu/mmp/kap18/problems/cd460.problem      0  NaN   
1400899               msu/mmp/kap18/problems/cd459.problem      0  NaN   
1400900               msu/mmp/kap18/problems/cd459.problem      0  NaN   

                      result  
0        incorrect_attempted  
1         correct_by_student  
2         correct_by_student  
3         correct_by_student  
4         correct_by_student  
...                      ...  
1400896  incorrect_attempted  
1400897   correct_by_student  
1400898   correct_by_student  
1400899   correct_by_student  
1400900  incorrect_attempted  

[1400901 rows x 7 columns]

In [263]:
# For homework questions, check how many students and how many questions in total
S_hw=hw_ques.to_numpy()[:,2]
N_hw=np.unique(S_hw)
print('The number of students N: ',len(N_hw))
# how many questions in exam
Q_hw=np.unique(hw_ques.to_numpy()[:,3])
print('The number of questions Q: ',len(Q_hw))
len(N_hw)
hw_ques_sort

The number of students N:  860
The number of questions Q:  161


timestamp         course        student  \
26847308  1211816944  178UapjfFklEc  42RyQy8XLRwNg   
26847306  1211816953  178UapjfFklEc  42RyQy8XLRwNg   
26846362  1217874374  178UapjfFklEc  42uHleBbtvATA   
26846360  1217874442  178UapjfFklEc  42uHleBbtvATA   
26846358  1217874450  178UapjfFklEc  42uHleBbtvATA   
...              ...            ...            ...   
25672432  1228990709  178UapjfFklEc  42irtwDSxtIYM   
25672430  1228990713  178UapjfFklEc  42irtwDSxtIYM   
25672425  1228990717  178UapjfFklEc  42irtwDSxtIYM   
25668941  1229003312  178UapjfFklEc  42hDZACMI.lLA   
25668920  1229003371  178UapjfFklEc  42hDZACMI.lLA   

                                                 resourceID partID mode  \
26847308  msu/kashy/physicsLib02/38_MW_Sound_Doppler/man...     11  NaN   
26847306  msu/kashy/physicsLib02/38_MW_Sound_Doppler/man...     11  NaN   
26846362  msu/kashy/Champ2002/champprobs/RectAreaPerimet...      0  NaN   
26846360  msu/kashy/Champ2002/champprobs/RectAreaPerimet...      0  NaN   
26846358  msu/kashy/Champ2002/champprobs/RectAreaPerimet...      0  NaN   
...                                                     ...    ...  ...   
25672432  msu/physicslib/msuphysicslib/01_Math_1/msu-pro...      0  NaN   
25672430  msu/physicslib/msuphysicslib/01_Math_1/msu-pro...      0  NaN   
25672425  msu/physicslib/msuphysicslib/01_Math_1/msu-pro...      0  NaN   
25668941                msu/nagytibo/Misc/presence2.problem      0  NaN   
25668920  msu/physicslib/msuphysicslib/01_Math_1/msu-pro...      0  NaN   

                       result  
26847308  incorrect_attempted  
26847306  incorrect_attempted  
26846362  incorrect_attempted  
26846360  incorrect_attempted  
26846358   correct_by_student  
...                       ...  
25672432  incorrect_attempted  
25672430  incorrect_attempted  
25672425  incorrect_attempted  
25668941   correct_by_student  
25668920   correct_by_student  

[409163 rows x 7 columns]

In [23]:
df_2=pd.read_csv('./adaptive-e-learning-for-educational-recommendation-system/DeepKnowledgeTracing-DKT-Pytorch/DKT/KTDataset/kddcup2010/kddcup2010_train.txt',header=None,sep='\t')
df_2
df_2_test=pd.read_csv('./adaptive-e-learning-for-educational-recommendation-system/DeepKnowledgeTracing-DKT-Pytorch/DKT/KTDataset/kddcup2010/kddcup2010_test.txt',header=None,sep='\t')
df_2_test
kddcup = pd.concat([df_2, df_2_test], axis=0)
kddcup

0
0                                                   436
1     447,568,580,394,41,91,643,152,474,523,428,526,...
2     1,1,1,1,1,1,0,0,1,1,1,1,1,1,1,0,0,1,1,0,0,0,1,...
3                                                   407
4     447,568,580,394,41,91,643,152,474,523,428,526,...
...                                                 ...
1609  447,568,580,394,41,91,643,152,474,523,428,526,...
1610  1,1,1,1,1,1,1,1,1,1,1,1,0,1,0,0,1,1,1,1,0,1,1,...
1611                                                 57
1612  447,568,580,394,41,91,643,152,474,523,428,526,...
1613  1,0,1,1,1,1,1,1,1,0,1,0,1,1,0,0,1,1,0,0,0,1,0,...

[8076 rows x 1 columns]

In [17]:
ques_dict_4 = {} # dictionary for question id
stud_dict_4 = {} # dictionary for student id
length=HW_ques_sort.to_numpy().shape[0]
comb_df_1=np.zeros((length,3))
Q_HW=
for i in range(len(Q_HW)):
    ques_dict_4[Q_HW[i]] = i  # You can assign values to the keys if needed
for i in range(len(N_HW)):
    stud_dict_4[N_HW[i]] = i  # You can assign values to the keys if needed
    
f1=np.frompyfunc(lambda hw_name, dic: dic[hw_name], 2, 1)

# # my_dict['msu/hallstei/PHY 231C/Homework/MISC/hw001.problem']
for n in range(length):
    stud=HW_ques_sort.to_numpy()[n,2]
    ques=HW_ques_sort.to_numpy()[n,3]
    result=HW_ques_sort.to_numpy()[n,6]
    q_id=f1(ques, ques_dict_4)
    s_id=f1(stud, stud_dict_4)
    comb_df_1[n,0]=s_id
    comb_df_1[n,1]=q_id
    if result=='correct_by_student':
        comb_df_1[n,2]=1
    else:
        comb_df_1[n,2]=0
comb_df_1=comb_df_1.astype(int)

In [20]:
file_path = "comb_df1_bkt.csv"
np.savetxt(file_path, comb_df_1, delimiter=',', fmt='%s')
comb_df_1=pd.read_csv('comb_df1_bkt.csv',header=None,sep='\t',delimiter=',')
comb_df_1
comb_df_1_train=comb_df_1.to_numpy()[0:1150000]
comb_df_1_test=comb_df_1.to_numpy()[1150000:]
np.savetxt('comb_df_1_train.csv', comb_df_1_train, delimiter=',', fmt='%s')
np.savetxt('comb_df_1_test.csv', comb_df_1_test, delimiter=',', fmt='%s')

In [315]:
my_dict_4 = {} # dictionary for question id

for i in range(len(Q_hw)):
    my_dict_4[Q_hw[i]] = i  # You can assign values to the keys if needed

# Print the dictionary
# print(my_dict)
f1=np.frompyfunc(lambda hw_name, dic: dic[hw_name], 2, 1)
# my_dict['msu/hallstei/PHY 231C/Homework/MISC/hw001.problem']
array_4=[]
for n in range(len(N_hw)):
    N=hw_ques[hw_ques['student']==N_hw[n]]
    array_4.append(N.shape[0]) # length
    q_list=N.to_numpy()[:,3][::-1]# we want order according to time
    result_list=N.to_numpy()[:,6][::-1]
    q_id=f1(q_list, my_dict_4)
    output = ''# string type
    for i in q_id:
        output += f"{i},"
    array_4.append(output) # ques
    ans=''
    for i in range(len(result_list)):
        if result_list[i]=='incorrect_attempted':
            ans +=str(0)+','
        elif result_list[i]=='correct_by_student':
            ans+=str(1)+','
    array_4.append(ans)
    

NameError: name 'total_array' is not defined

In [145]:
file_path = "course27.csv"

# Save the NumPy array as a CSV file
np.savetxt(file_path, array_4, delimiter=',', fmt='%s')

In [187]:
course27_train[7]

array(['32,67,5,68,68,68,68,68,68,68,65,65,65,31,68,68,68,68,68,68,65,65,68,68,68,68,68,68,69,69,69,69,69,69,70,71,71,71,71,71,72,72,78,6,73,72,77,77,71,78,78,74,74,74,27,27,26,26,72,72,25,25,27,27,26,75,75,76,76,76,76,76,76,76,79,79,79,79,79,81,156,156,86,86,86,86,76,76,79,80,82,83,85,85,85,85,85,157,157,157,157,86,86,89,89,76,76,76,84,157,76,157,157,157,157,157,157,157,86,86,86,89,89,158,159,159,160,160,161,161,161,161,162,162,162,162,162,162,162,163,163,163,163,164,164,164,164,164,164,164,93,93,164,16,16,16,165,166,166,166,92,92,92,92,92,92,92,92,92,92,93,93,93,0,0,141,142,142,146,140,21,1,13,18,18,148,148,148,167,167,22,22,22,22,168,168,169,169,169,170,94,94,91,95,170,170,94,94,94,94,94,94,94,96,96,96,10,87,87,170,170,171,171,171,94,94,94,94,171,170,97,7,7,171,171,171,7,170,88,97,7,170,170,96,96,96,96,97,87,87,170,87,87,94,177,177,177,177,177,177,177,177,177,177,100,103,103,101,172,98,98,176,173,174,174,174,174,174,180,180,180,180,99,104,104,175,175,179,179,179,105,105,105,105,105,

In [202]:
course4_train=course4.to_numpy()[0:2100]
course4_test=course4.to_numpy()[2100:]
np.savetxt('course4_train.csv', course4_train, delimiter=',', fmt='%s')
np.savetxt('course4_test.csv', course4_test, delimiter=',', fmt='%s')
# train_27=course27.iloc[:1500]
# test_27=course27.iloc[1500:]
#course27_train.to_csv("course27_train.csv")
#test_27.to_csv("test_27.csv")

In [201]:
course4=pd.read_csv('course4.csv',header=None,sep='\t')
course4

0
0                                                   540
1     56,75,75,75,76,2,2,77,77,77,1,1,1,3,11,79,80,8...
2     1,0,0,1,1,0,1,0,0,1,0,0,1,1,1,1,0,0,0,0,0,0,0,...
3                                                   501
4     56,75,76,2,2,77,77,77,77,77,77,77,77,77,77,77,...
...                                                 ...
2575  1,1,1,3,3,3,3,3,11,11,11,11,11,11,79,80,80,80,...
2576  0,0,1,0,0,0,0,1,0,0,0,0,0,1,1,0,0,0,1,0,0,0,0,...
2577                                                680
2578  56,75,1,1,1,76,75,79,79,3,11,11,79,80,80,80,81...
2579  1,0,0,0,1,1,1,0,0,1,0,1,1,0,0,1,0,0,0,0,0,0,0,...

[2580 rows x 1 columns]

In [250]:
# N0=hw_ques[hw_ques['student']==N_hw[2]]
# print(np.unique(N0.to_numpy()[:,3]).shape)
# print(N0.to_numpy()[:,3].shape)

# check if the order is based on timestamp
# and print out how many unique exercise each studend takes
count=0
Q_uni=[]
idx=[]
hw_ques_copy=hw_ques.copy()
for n in range(len(N_hw)):
    N=hw_ques[hw_ques['student']==N_hw[n]]
    sorted_df = N.sort_values(by='timestamp', ascending=False)
#    sorted_df = sorted_df.reset_index(drop=True)
#    hw_ques.loc[sorted_df.index] = sorted_df
#    hw_ques.loc[hw_ques['student']==N_hw[n]]=sorted_df.values
    # Check if the sorted order m#atches the original order
    is_order=N.equals(sorted_df)
    if is_order:
        count+=1
    else:
        print('the unordered student is:',n)
        idx.append(n)
    Q_uni.append(np.unique(N.to_numpy()[:,3]).shape[0])
print(count)
print('How many unique questions each student takes:')
print(Q_uni)
np.unique(Q_uni)

the unordered student is: 33
the unordered student is: 43
the unordered student is: 46
the unordered student is: 59
the unordered student is: 115
the unordered student is: 135
the unordered student is: 146
the unordered student is: 193
the unordered student is: 236
the unordered student is: 276
the unordered student is: 287
the unordered student is: 321
the unordered student is: 341
the unordered student is: 411
the unordered student is: 425
the unordered student is: 447
the unordered student is: 462
the unordered student is: 463
the unordered student is: 558
the unordered student is: 609
the unordered student is: 610
the unordered student is: 611
the unordered student is: 661
the unordered student is: 712
the unordered student is: 859
835
How many unique questions each student takes:
[140, 144, 144, 142, 136, 132, 145, 18, 142, 145, 139, 126, 144, 33, 145, 140, 13, 135, 145, 42, 18, 27, 139, 27, 124, 128, 145, 142, 141, 145, 143, 138, 12, 145, 130, 20, 144, 140, 143, 32, 138, 145, 128

array([  1,   2,   4,   5,   6,   7,   8,  11,  12,  13,  15,  18,  19,
        20,  21,  22,  24,  26,  27,  28,  29,  30,  32,  33,  34,  37,
        38,  40,  42,  45,  48,  50,  52,  55,  56,  58,  59,  60,  61,
        62,  63,  65,  66,  68,  71,  74,  77,  79,  80,  83,  84,  89,
        91,  92,  93,  94,  95,  96,  97,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 151, 152, 156, 157])

In [253]:
N_hw_arr=[] # show each student takes how many homework questions
for i in range(len(N_hw)):
    N1=hw_ques[hw_ques['student']==N_hw[i]]
    N_hw_arr.append(N1.shape[0])
print('How many TOTAL questions each student takes:',end='\n')
print(N_hw_arr)
N1

How many TOTAL questions each student takes:
[540, 501, 466, 410, 433, 521, 540, 73, 502, 488, 686, 356, 560, 119, 493, 545, 39, 563, 503, 124, 62, 62, 736, 70, 403, 372, 576, 391, 647, 491, 474, 577, 69, 596, 563, 87, 514, 467, 657, 101, 709, 522, 732, 681, 627, 153, 680, 631, 455, 76, 432, 509, 414, 479, 652, 551, 437, 526, 540, 545, 119, 332, 427, 150, 527, 704, 76, 536, 554, 443, 636, 449, 455, 622, 492, 596, 342, 513, 104, 411, 439, 259, 348, 404, 450, 406, 370, 620, 446, 36, 564, 31, 724, 582, 602, 445, 356, 621, 414, 749, 873, 477, 415, 332, 453, 370, 484, 445, 497, 505, 534, 667, 414, 506, 434, 617, 542, 810, 212, 477, 274, 429, 569, 549, 482, 479, 553, 467, 465, 423, 398, 392, 267, 495, 448, 475, 443, 552, 471, 583, 494, 428, 563, 447, 419, 461, 524, 463, 140, 53, 560, 443, 515, 360, 634, 757, 21, 535, 621, 497, 715, 489, 442, 615, 466, 960, 479, 776, 435, 566, 433, 649, 324, 466, 246, 531, 430, 364, 88, 618, 498, 737, 657, 472, 455, 727, 422, 450, 488, 467, 562, 387, 493, 676

timestamp         course        student  \
25717185  1228420681  178UapjfFklEc  42zqlZ75j5aAg   
25717275  1228420481  178UapjfFklEc  42zqlZ75j5aAg   
25717326  1228420374  178UapjfFklEc  42zqlZ75j5aAg   
25717393  1228420231  178UapjfFklEc  42zqlZ75j5aAg   
25719437  1228415865  178UapjfFklEc  42zqlZ75j5aAg   
...              ...            ...            ...   
26826334  1219869515  178UapjfFklEc  42zqlZ75j5aAg   
26826351  1219869349  178UapjfFklEc  42zqlZ75j5aAg   
26826352  1219869339  178UapjfFklEc  42zqlZ75j5aAg   
26835220  1219776797  178UapjfFklEc  42zqlZ75j5aAg   
26835240  1219776718  178UapjfFklEc  42zqlZ75j5aAg   

                                                 resourceID partID mode  \
25717185               msu/mmp/kap14/problems/cd413.problem      0  NaN   
25717275               msu/mmp/kap14/problems/cd413.problem      0  NaN   
25717326               msu/mmp/kap14/problems/cd418.problem      0  NaN   
25717393              msu/mmp/kap14/problems/cd418c.problem      0  NaN   
25719437               msu/mmp/kap14/problems/cd416.problem      0  NaN   
...                                                     ...    ...  ...   
26826334  msu/kashy/Champ2002/champprobs/RectAreaPerimet...      0  NaN   
26826351  msu/kashy/Champ2002/champprobs/RectAreaPerimet...      0  NaN   
26826352  msu/kashy/Champ2002/champprobs/RectAreaPerimet...      0  NaN   
26835220  msu/physicslib/msuphysicslib/01_Math_1/msu-pro...      0  NaN   
26835240                msu/nagytibo/Misc/presence2.problem      0  NaN   

                       result  
25717185  incorrect_attempted  
25717275  incorrect_attempted  
25717326   correct_by_student  
25717393   correct_by_student  
25719437   correct_by_student  
...                       ...  
26826334   correct_by_student  
26826351  incorrect_attempted  
26826352  incorrect_attempted  
26835220  incorrect_attempted  
26835240   correct_by_student  

[680 rows x 7 columns]

Dataset for course 0

In [400]:
# how many students in exam
s1=exam1.to_numpy()[:,2]
N=np.unique(s1) # how many students taking exam
print('There are',len(N),'students taking exam of this course')

N1_arr=[] # show each student takes how many exam questions
for i in range(len(N)):
    N1=exam1[exam1['student']==N[i]]
    N1_arr.append(N1.shape[0])
print('The number of questions taken by each student is as follow:',end='\n')
print(N1_arr)

There are 461 students taking exam of this course
The number of questions taken by each student is as follow:
[34, 18, 34, 34, 34, 34, 34, 34, 9, 18, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 33, 34, 34, 33, 34, 34, 34, 34, 34, 34, 34, 34, 33, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 30, 34, 34, 34, 34, 33, 34, 33, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 33, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 33, 34, 34, 34, 34, 34, 34, 33, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 25, 34, 34, 9, 34, 34, 34, 34, 34, 34, 34, 34, 25, 18, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 33, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 33, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 34, 33, 34, 34, 34, 34,

In [402]:
# Use dataset where students take 34 questions
indices = [index for index, value in enumerate(N1_arr) if value <= 34]
N_real=N[indices]
# student=exam1[exam1['student']==N_real[105]]
# ques_curr=student[student['resourceID']==Q_new[17]].to_numpy()
# ques_curr
N_real.shape

# if ques_curr[:,6]=='correct_by_scantron':
#     print(1)

(460,)

In [403]:
# only consider studnets who take 33 questions
# check if these questions are the same among all students
question=np.unique(exam1[exam1['student']==N_real[0]].to_numpy()[:,3])
question
count=0
for i in range(len(N_real)):
    
    question2=np.unique(exam1[exam1['student']==N_real[i]].to_numpy()[:,3])
    if (question==question2).all():
        count=count+1
print(count)

## Yes, they all match

<ipython-input-403-4297df45b46e>:9: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  if (question==question2).all():


AttributeError: 'bool' object has no attribute 'all'

In [373]:
Q_new.shape

(31,)

In [370]:
N1=exam1[exam1['student']==N_real[0]]

for n in range(1,len(N_real)):
    N1=N1.append(exam1[exam1['student']==N_real[n]])

In [371]:
N1

timestamp         course        student  \
1979703  1134567414  17oPA6FLPK2B.  42.GIof9MbEZA   
1979709  1134567413  17oPA6FLPK2B.  42.GIof9MbEZA   
1979712  1134567413  17oPA6FLPK2B.  42.GIof9MbEZA   
1979713  1134567413  17oPA6FLPK2B.  42.GIof9MbEZA   
1979716  1134567412  17oPA6FLPK2B.  42.GIof9MbEZA   
...             ...            ...            ...   
2841241  1126542995  17oPA6FLPK2B.  42zaEAai8fr/w   
2841242  1126542995  17oPA6FLPK2B.  42zaEAai8fr/w   
2841244  1126542995  17oPA6FLPK2B.  42zaEAai8fr/w   
2841247  1126542995  17oPA6FLPK2B.  42zaEAai8fr/w   
2841251  1126542994  17oPA6FLPK2B.  42zaEAai8fr/w   

                                                resourceID partID  mode  \
1979703  msu/kortemey/physicslib/oscillations/horizonta...     12  exam   
1979709  msu/kortemey/physicslib/rotation/startrun.problem      0  exam   
1979712  msu/kortemey/physicslib/two_dim_motion/xvatf.p...      0  exam   
1979713  msu/kortemey/physicslib/two_dim_motion/graphso...      0  exam   
1979716  msu/kortemey/physicslib/momentum/truefalseradi...      0  exam   
...                                                    ...    ...   ...   
2841241  msu/kortemey/physicslib/units/dimensionality.p...      0  exam   
2841242  msu/kortemey/physicslib/one_dim_motion/positio...     15  exam   
2841244  msu/kortemey/physicslib/two_dim_motion/takeoff...      0  exam   
2841247       msu/kortemey/physicslib/units/newton.problem      0  exam   
2841251  msu/kortemey/physicslib/two_dim_motion/graphso...      0  exam   

                      result  
1979703  correct_by_scantron  
1979709  incorrect_attempted  
1979712  correct_by_scantron  
1979713  incorrect_attempted  
1979716  incorrect_attempted  
...                      ...  
2841241  correct_by_scantron  
2841242  incorrect_attempted  
2841244  correct_by_scantron  
2841247  correct_by_scantron  
2841251  correct_by_scantron  

[16228 rows x 7 columns]

In [215]:
# find out one repeated element
Q1=exam1[exam1['student']==N_real[1]].to_numpy()[:,3]
Q1
unique_elements, counts = np.unique(Q1, return_counts=True)
repeated_elements = unique_elements[counts > 1]
print("Repeated elements:", repeated_elements)
# every student took this repeated question twice, so we drop this question from our set
Q_new = Q[Q != repeated_elements]
Q_new.shape

Repeated elements: ['msu/physicslib/msuphysicslib/64_EMInduction1/msuprob24b.problem']


(31,)

In [261]:
Y
excel_filename = "Y_binary.xlsx"

Y_binary = pd.DataFrame(Y)
Y_binary.to_excel(excel_filename, index=False, header=False)

In [258]:
Y=np.zeros((len(Q_new),len(N_real)))
Y.shape
for i in range(len(Q_new)):
    for j in range(len(N_real)):
        student_curr=exam1[exam1['student']==N_real[j]]
        ques_curr=student_curr[student_curr['resourceID']==Q_new[i]].to_numpy()
        if ques_curr[:,6]=='correct_by_scantron':
            Y[i,j]=1
        elif ques_curr[:,6]=='incorrect_attempted':
            Y[i,j]=0
        else: 
            Y[i,j]=None

In [374]:
Y.shape

(31, 297)

# Hand mode

In [126]:
# how many students in hand mode
h1=hand1.to_numpy()[:,2]
N_hand=np.unique(h1)
print('The number of students N: ',len(N_hand))
# how many questions in hand mode
hq1=exam1.to_numpy()[:,3]
Q_hand=np.unique(hq1)
print('The number of questions Q: ',len(Q_hand))

The number of students N:  352
The number of questions Q:  32


In [127]:
hand1

timestamp         course        student  \
14739627  1228921162  170Rg65URrGNE  42JOsQRD35K5g   
14739628  1228921161  170Rg65URrGNE  42JOsQRD35K5g   
14739629  1228921161  170Rg65URrGNE  42JOsQRD35K5g   
14739630  1228921161  170Rg65URrGNE  42JOsQRD35K5g   
14739631  1228921161  170Rg65URrGNE  42JOsQRD35K5g   
...              ...            ...            ...   
15381116  1221433785  170Rg65URrGNE  4256Y2Nez60C2   
15381119  1221433785  170Rg65URrGNE  42hbPUT8oe4RA   
15381122  1221433785  170Rg65URrGNE  42e7IZ6IT.72o   
15381123  1221433785  170Rg65URrGNE  42DF2VHgVEio6   
15409081  1220949263  170Rg65URrGNE  42q9Sc10cdAWs   

                                                 resourceID partID  mode  \
14739627  wiley/cutnell/Physics_7ed/Chap28/cj-prob2842.p...      0  hand   
14739628  georgetown/skelton/Phys-042/PLQs/PLQ19/Negativ...      0  hand   
14739629  brookscole/serway/College_Physics_6ed/Chap23/p...      0  hand   
14739630  brookscole/serway/College_Physics_6ed/Chap24/p...      0  hand   
14739631  georgetown/skelton/Phys-042/PLQs/PLQ20/Fringes...      0  hand   
...                                                     ...    ...   ...   
15381116     msu/browned/PHY184/HW3/Charged-Spheres.problem    NaN  hand   
15381119     msu/browned/PHY184/HW3/Charged-Spheres.problem     17  hand   
15381122     msu/browned/PHY184/HW3/Charged-Spheres.problem     17  hand   
15381123     msu/browned/PHY184/HW3/Charged-Spheres.problem    NaN  hand   
15409081              msu/mmp/kap18/problems/cd460a.problem      0  hand   

                         result  
14739627    correct_by_override  
14739628    correct_by_override  
14739629  incorrect_by_override  
14739630  incorrect_by_override  
14739631    correct_by_override  
...                         ...  
15381116  incorrect_by_override  
15381119  incorrect_by_override  
15381122  incorrect_by_override  
15381123  incorrect_by_override  
15409081    correct_by_override  

[655 rows x 7 columns]

In [128]:
# how many students in hand
h1=hand1.to_numpy()[:,2]
N_hand=np.unique(h1) # how many students taking exam
print('There are',len(N_hand),'students taking exam of this course')

N1_hand=[] # show each student takes how many exam questions
for i in range(len(N_hand)):
    N1=hand1[hand1['student']==N_hand[i]]
    N1_hand.append(N1.shape[0])
print('The number of questions taken by each student is as follow:',end='\n')
print(N1_hand)

There are 352 students taking exam of this course
The number of questions taken by each student is as follow:
[1, 1, 1, 1, 1, 3, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 20, 1, 1, 1, 1, 1, 1, 2, 9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 11, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 11, 1, 9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 20, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 20, 1, 6, 7, 1, 1, 20, 1, 2, 1, 2, 3, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 3, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 20, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 10, 20, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 9, 1, 9, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 7, 8, 1, 1

In [118]:
hand1[hand1['student']==N_hand[1]]

timestamp         course        student  \
16718990  1202746274  173XHazC4puKI  421n1Yp1qMLq.   
16718991  1202746274  173XHazC4puKI  421n1Yp1qMLq.   
16718992  1202746274  173XHazC4puKI  421n1Yp1qMLq.   
16718993  1202746274  173XHazC4puKI  421n1Yp1qMLq.   
16718994  1202746274  173XHazC4puKI  421n1Yp1qMLq.   
16718995  1202746273  173XHazC4puKI  421n1Yp1qMLq.   
16718996  1202746273  173XHazC4puKI  421n1Yp1qMLq.   
16718997  1202746273  173XHazC4puKI  421n1Yp1qMLq.   
16718998  1202746273  173XHazC4puKI  421n1Yp1qMLq.   
16718999  1202746273  173XHazC4puKI  421n1Yp1qMLq.   

                                                 resourceID partID  mode  \
16718990  fsu/capalibrary/54ElectricPotential1/prob03.pr...      0  hand   
16718991  csm/csmphyslib/type54_potential/ParallelPlate....      0  hand   
16718992  brookscole/serway/College_Physics_6ed/Chap16/p...      0  hand   
16718993  fsu/capalibrary/56CapacitanceCapacitors/prob03...      0  hand   
16718994  msu/physicslib/msuphysicslib/56_Capacitance/ms...      0  hand   
16718995  msu/schwier/phy184_ss2008/directionOfForceNega...      0  hand   
16718996           ohiou/OUp200lib/sfChpt15/sf1516a.problem      0  hand   
16718997  fsu/giancolilib/Chap16_ElectricChargeField/gcp...      0  hand   
16718998                 msu/stump/PHY184/Ex1/Wk2-3.problem      0  hand   
16718999                 msu/stump/PHY184/Ex1/Wk3_2.problem      0  hand   

                         result  
16718990    correct_by_override  
16718991    correct_by_override  
16718992    correct_by_override  
16718993    correct_by_override  
16718994    correct_by_override  
16718995  incorrect_by_override  
16718996    correct_by_override  
16718997  incorrect_by_override  
16718998  incorrect_by_override  
16718999    correct_by_override

In [242]:
# was there a wrong datalog?
N_59[N_59['resourceID']==Q59[0]]

timestamp         course        student  \
14973166  1225459548  170Rg65URrGNE  428IPdmipoH6U   
14973184  1225459544  170Rg65URrGNE  428IPdmipoH6U   
14973198  1225459541  170Rg65URrGNE  428IPdmipoH6U   
14973215  1225459536  170Rg65URrGNE  428IPdmipoH6U   
14973230  1225459531  170Rg65URrGNE  428IPdmipoH6U   
...              ...            ...            ...   
14978238  1225458016  170Rg65URrGNE  428IPdmipoH6U   
14978256  1225458012  170Rg65URrGNE  428IPdmipoH6U   
14978271  1225458007  170Rg65URrGNE  428IPdmipoH6U   
14978287  1225458003  170Rg65URrGNE  428IPdmipoH6U   
14978303  1225457999  170Rg65URrGNE  428IPdmipoH6U   

                                                 resourceID partID  mode  \
14973166  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14973184  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14973198  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14973215  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14973230  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
...                                                     ...    ...   ...   
14978238  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14978256  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14978271  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14978287  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   
14978303  msu/nagytibo/Electromagnetism/DCcircuits/011a....    NaN  exam   

                       result  
14973166  noanswer_or_invalid  
14973184  noanswer_or_invalid  
14973198  noanswer_or_invalid  
14973215  noanswer_or_invalid  
14973230  noanswer_or_invalid  
...                       ...  
14978238  noanswer_or_invalid  
14978256  noanswer_or_invalid  
14978271  noanswer_or_invalid  
14978287  noanswer_or_invalid  
14978303  noanswer_or_invalid  

[321 rows x 7 columns]

In [241]:
#outlier
N_59=exam1[exam1['student']==N[59]]
Q_59=N_59.to_numpy()[:,3]
Q59=np.unique(Q_59)
Q59.shape

(14,)

In [179]:
N1[N1.duplicated(['resourceID'],keep='last')]

Empty DataFrame
Columns: [timestamp, course, student, resourceID, partID, mode, result]
Index: []

In [178]:
s1.shape

(21049,)

In [62]:

for i, arr in enumerate(array_group):
    csv_name = f'sheet{i+1}.csv'  # Sheet names will be Sheet1, Sheet2, ..., Sheet34
    df_course = pd.DataFrame(arr)
    df_course.to_csv(csv_name,index=False)
